interface/src/checksums_storage.rs

```diff
@@ -0,0 +1,149 @@
+use async_trait::async_trait;
+use solana_sdk::pubkey::Pubkey;
+
+/// Type of checksum for bubblegum epochs and account NFT buckets.
+/// It is technically a SHA3 hash.
+pub type Chksm = [u8; 32];
+
+/// Data transfer object for bubblegum grand epoch.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmGrandEpochCksm {
+    pub tree_pubkey: Pubkey,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for bubblegum grand epoch checksums
+/// for listing multiple grand epochs for a tree
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmGrandEpochCksmWithNumber {
+    pub grand_epoch: u16,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for bubblegum epoch.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmEpochCksm {
+    pub epoch: u32,
+    pub tree_pubkey: Pubkey,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for bubblegum tree change.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct BbgmChangeRecord {
+    pub tree_pubkey: Pubkey,
+    pub slot: u64,
+    pub seq: u64,
+    pub signature: String,
+}
+
+/// Used to specify offset when fetching a portion of
+/// bubblegum tree changes.
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmChangePos {
+    pub slot: u64,
+    pub seq: u64,
+}
+
+/// Data transfer object for account NFT grand bucket.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AccGrandBucketCksm {
+    pub grand_bucket: u16,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for account NFT bucket.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AccBucketCksm {
+    pub bucket: u16,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for account NFT last received state.
+#[derive(Debug, Clone, PartialOrd, Ord)]
+pub struct AccLastChange {
+    pub account_pubkey: Pubkey,
+    pub slot: u64,
+    pub write_version: u64,
+    pub data_hash: u64,
```

Comment:
As far as we are tracking not only slot+write_version for accounts but also data_hash, we need to discuss the consensus mechanism. 

If i understood correctly, the current implementation would identify accounts with the same slot+write_version but different data_hashes as different updates and will try to synchronize them between nodes. However the current accounts processing mechanism will not allow updating existing accounts with the same  slot+write_version. So nodes will try to receive "updates" from each other, send these updates to each other but not process them correctly. 

Maybe this sounds like a separate task but we need to discuss approach that we think will be the better

Reply:
This is purely for account NFTs, because for account we don't have cleaning of forks, and as the result we cannot rely on slot. That's why we ignore slots and take only `pubkey+write_version_data_hash` when we calculate account changes checksum, and when we compare peer's account change with local account change.
If the `slot` is different, but the `data_hash` is same, we assume that's the same account change.

Reply:
For the consensus part - that's for us to define. Essentially, we should expect the following:
- same write_version, same slot, same hash - all good
- same write_version, same slot, different hash - we need the consensus to define, what the correct version is. I'd suggest we ask RPC of the current state (if it has not changed yet, of course) 
- same write_version, different slot, same hash - all good, take the highest slot on merge probably
- same write_version, different slot, different hashes - looks like a fork, take the highest slot on merge
- different write_version, same/different slot, same/different hash - take the highest write_version